src/borg/archiver.py

```diff
@@ -1222,6 +1226,91 @@ def do_debug_dump_archive_items(self, args, repository, manifest, key):
         print('Done.')
         return EXIT_SUCCESS
 
+    def debug_json(self, d):
+        def decode_bytes(value):
+            # this should somehow be reversable later, but usual strings should
+            # look nice and chunk ids should mostly show in hex. Use a special
+            # inband signaling character (ASCII DEL) to distinguish between
+            # decoded and hex mode.
+            if not value.startswith(b'\x7f'):
+                try:
+                    value = value.decode()
+                    return value
+                except UnicodeDecodeError:
+                    pass
+            return '\u007f' + bin_to_hex(value)
+
+        def decode_tuple(t):
+            res = []
+            for value in t:
+                if isinstance(value, dict):
+                    value = decode(value)
+                elif isinstance(value, tuple) or isinstance(value, list):
+                    value = decode_tuple(value)
+                elif isinstance(value, bytes):
+                    value = decode_bytes(value)
+                res.append(value)
+            return res
+
+        def decode(d):
+            res = collections.OrderedDict()
+            for key, value in d.items():
+                if isinstance(value, dict):
+                    value = decode(value)
+                elif isinstance(value, tuple) or isinstance(value, list):
+                    value = decode_tuple(value)
+                elif isinstance(value, bytes):
+                    value = decode_bytes(value)
+                if isinstance(key, bytes):
+                    res[key.decode()] = value
+                else:
+                    res[key] = value
+            return res
+
+        return decode(d)
+
+    @with_repository()
+    def do_debug_dump_archive(self, args, repository, manifest, key):
+        """dump decoded archive metadata (not: data)"""
+
+        try:
+            archive_meta_orig = manifest.archives.get_raw_dict()[safe_encode(args.location.archive)]
+        except KeyError:
+            raise Archive.DoesNotExist(args.location.archive)
+        archive_meta = self.debug_json(archive_meta_orig)
+        archive_meta['_name'] = args.location.archive
+
+        _, data = key.decrypt(archive_meta_orig[b'id'], repository.get(archive_meta_orig[b'id']))
+        archive_org_dict = msgpack.unpackb(data, object_hook=StableDict, unicode_errors='surrogateescape')
+        archive_meta['_meta'] = self.debug_json(archive_org_dict)
+
+        unpacker = msgpack.Unpacker(use_list=False, object_hook=StableDict)
+        items = []
+        for item_id in archive_org_dict[b'items']:
+            _, data = key.decrypt(item_id, repository.get(item_id))
+            unpacker.feed(data)
+            for item in unpacker:
+                item = self.debug_json(item)
+                items.append(item)
+
+        archive_meta['_items'] = items
+
+        with open(args.path, 'w') as fd:
```

Comment:
instead of explicitly giving /dev/stdout, this should be the default if not given.